pkg/leader/leader.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2018 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package leader
+
+import (
+	"context"
+	"errors"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/sirupsen/logrus"
+)
+
+// ErrNoNS indicates that a namespace could not be found for the current
+// environment
+var ErrNoNS = errors.New("namespace not found for current environment")
+
+// TryBecome behaves like Become, except it will not return an error in the
+// case where a namespace cannot be found for the current pod. This is useful
+// for a service that might run outside the cluster, for example an operator
+// being started with `operator-sdk up local`.
+func TryBecome(ctx context.Context, name string) error {
```

Comment:
It maybe more intuitive if we call `name` to either `lockKey` or `leaderKey`.

Reply:
I was thinking `name` because the value will literally go into the `name` field on the `ConfigMap`. Maybe `lockName` would be better?

I worry about using the term "key" in a scenario that involves a "lock", since we don't want anyone to think that it is the kind of key that could "unlock" something.

Thoughts on that? I'm open to suggestions.

Reply:
`lockName` sounds better. The reason I think having a keyword `lock` is useful because the usr now expects multiple instances having the same `lockName` going compete and one will win.